File: src/lib/ai/providers/opencodeProvider.ts
const OPENCODE_API_URL = (process.env.OPENCODE_API_URL ?? "https://ttt.voidall.com").replace(/\/$/, "");

// Sent as X-API-Key on every call to either TTT backend (OpenCode and OCR
// share one key); they reject requests without it.
export function tttAuthHeaders(): Record<string, string> {
  const key = process.env.TTT_API_KEY;
  return key ? { "X-API-Key": key } : {};
}

export async function opencodeText(prompt: string, system: string): Promise<string> {
  const submitRes = await fetch(`${OPENCODE_API_URL}/api/tasks/upload`, {
    method: "POST",
    headers: { "Content-Type": "application/json", ...tttAuthHeaders() },
    body: JSON.stringify({ text: prompt, system_prompt: system, model: "opencode" }),
  });
  if (!submitRes.ok) throw new Error(`OpenCode submit failed: ${submitRes.status}`);
  const { id } = await submitRes.json();

  const deadline = Date.now() + 420_000;
  while (Date.now() < deadline) {
    await new Promise((r) => setTimeout(r, 3000));
    const pollRes = await fetch(`${OPENCODE_API_URL}/api/tasks/${id}`, {
      headers: tttAuthHeaders(),
    });
    if (!pollRes.ok) throw new Error(`OpenCode poll failed: ${pollRes.status}`);
    const task = await pollRes.json();
    if (task.status === "completed") {
      try {
        const result = JSON.parse(task.result as string) as { response?: string };
        return result.response ?? "";
      } catch {
        throw new Error("OpenCode returned invalid JSON");
      }
    }
    if (task.status === "failed") throw new Error(`OpenCode task failed: ${task.error}`);
  }
  throw new Error("OpenCode task timed out after 420s");
}

File: backend/app/ai/providers/opencode_provider.py
"""OpenCode provider — port of src/lib/ai/providers/opencodeProvider.ts."""
import asyncio
import json
import time

import httpx

from app.config import settings


def _base_url() -> str:
    return (settings.opencode_api_url or "https://opencode.voidall.com").rstrip("/")


def auth_headers() -> dict[str, str]:
    """X-API-Key for both TTT backends (OpenCode and OCR share one key)."""
    return {"X-API-Key": settings.ttt_api_key} if settings.ttt_api_key else {}


async def opencode_text(prompt: str, system: str) -> str:
    base = _base_url()
    async with httpx.AsyncClient(timeout=30.0, headers=auth_headers()) as client:
        submit = await client.post(
            f"{base}/api/tasks/upload",
            json={"text": prompt, "system_prompt": system, "model": "opencode"},
        )
        if submit.status_code >= 400:
            raise RuntimeError(f"OpenCode submit failed: {submit.status_code}")
        task_id = submit.json()["id"]

        deadline = time.monotonic() + 420.0
        while time.monotonic() < deadline:
            await asyncio.sleep(3.0)
            poll = await client.get(f"{base}/api/tasks/{task_id}")
            if poll.status_code >= 400:
                raise RuntimeError(f"OpenCode poll failed: {poll.status_code}")
            task = poll.json()
            if task.get("status") == "completed":
                try:
                    result = json.loads(task.get("result") or "")
                except (json.JSONDecodeError, TypeError):
                    raise RuntimeError("OpenCode returned invalid JSON")
                return result.get("response") or ""
            if task.get("status") == "failed":
                raise RuntimeError(f"OpenCode task failed: {task.get('error')}")
    raise RuntimeError("OpenCode task timed out after 420s")
